File: microcol-game/src/main/java/org/microcol/model/Location.java
package org.microcol.model;

import java.util.List;

import com.google.common.base.MoreObjects;
import com.google.common.base.Preconditions;
import com.google.common.collect.ImmutableList;

public final class Location {

    /**
     * Minimal map x-axe value.
     */
    public static final int MAP_MIN_X = 1;

    /**
     * Minimal map y-axe value.
     */
    public static final int MAP_MIN_Y = 1;

    private final int x;
    private final int y;

    private Location(final int x, final int y) {
        this.x = x;
        this.y = y;
    }

    public int getX() {
        return x;
    }

    public int getY() {
        return y;
    }

    /**
     * Return distance between two locations. It use Euclidean distance.
     *
     * @param location
     *            required location
     * @return distance between two locations
     */
    public int getDistance(final Location location) {
        Preconditions.checkNotNull(location);

        final int diffX = x - location.x;
        final int diffY = y - location.y;
        return (int) Math.round(Math.sqrt(diffX * diffX + diffY * diffY));
    }

    /**
     * Return distance between two locations. It use Manhattan distance. It's
     * described at <a href= "https://en.wiktionary.org/wiki/Manhattan_distance"
     * >https://en.wiktionary.org/wiki/Manhattan_distance</a>.
     * <p>
     * This distance is better for path calculations because it nicely work with
     * integers.
     * </p>
     *
     * @param location
     *            required location
     * @return distance between two locations
     */
    public int getDistanceManhattan(final Location location) {
        Preconditions.checkNotNull(location);

        return Math.abs(x - location.x) + Math.abs(y - location.y);
    }

    public Location add(final Location location) {
        Preconditions.checkNotNull(location);

        return Location.of(x + location.x, y + location.y);
    }

    public Location sub(final Location location) {
        Preconditions.checkNotNull(location);

        return Location.of(x - location.x, y - location.y);
    }

    public boolean isNeighbor(final Location location) {
        Preconditions.checkNotNull(location);

        if (equals(location)) {
            return false;
        }

        return Math.abs(x - location.x) <= 1 && Math.abs(y - location.y) <= 1;
    }

    public List<Location> getNeighbors() {
        return Direction.getVectors().stream().map(direction -> add(direction))
                .collect(ImmutableList.toImmutableList());
    }

    @Override
    public int hashCode() {
        return x + (y << Integer.SIZE / 2);
    }

    @Override
    public boolean equals(final Object object) {
        if (object == null) {
            return false;
        }

        if (!(object instanceof Location)) {
            return false;
        }

        final Location location = (Location) object;

        return x == location.x && y == location.y;
    }

    @Override
    public String toString() {
        return MoreObjects.toStringHelper(this).add("x", x).add("y", y).toString();
    }

    public boolean isDirection() {
        return Direction.getVectors().contains(this);
    }

    public static Location of(final int x, final int y) {
        return new Location(x, y);
    }
}

File: microcol-game/src/main/kotlin/org/microcol/model/Location.kt
package org.microcol.model

import kotlin.math.abs
import kotlin.math.round
import kotlin.math.sqrt

data class Location private constructor(val x: Int, val y: Int) {
    operator fun plus(location: Location) = Location.of(x + location.x, y + location.y)

    // Tohle tu nechávám pouze protože to je v API a nechce se mi to teď všude měnit.
    fun add(location: Location) = this + location

    // WTF CO TO JE?
    operator fun minus(location: Location) = Location.of(x - location.x, y - location.y)

    // Tohle tu nechávám pouze protože to je v API a nechce se mi to teď všude měnit.
    fun sub(location: Location) = this - location

    fun isNeighbor(location: Location) = if (equals(location)) false else abs(x - location.x) <= 1 && abs(y - location.y) <= 1

    val neighbors
        get() = Direction.getVectors().map { direction -> add(direction) }

    // WTF CO TO JE?
    val isDirection
        get() = Direction.getVectors().contains(this)

    fun getDistance(location: Location): Int {
        val (diffX, diffY) = this - location
        val tmp = (diffX * diffX + diffY * diffY).toDouble()

        return round(sqrt(tmp)).toInt()
    }

    fun getDistanceManhattan(location: Location) = abs(x - location.x) + abs(y - location.y)

    override fun hashCode() = x + (y shl Integer.SIZE / 2)

    // TAKHLE SE DĚLAJÍ STATICKÉ METODY
    companion object {
        // TOHLE TU VŮBEC NEMÁ BÝT
        const val MAP_MIN_X = 1

        // TOHLE TU VŮBEC NEMÁ BÝT
        const val MAP_MIN_Y = 1

        @JvmStatic
        fun of(x: Int, y: Int): Location {
            return Location(x, y)
        }
    }
}
